packages/localize-tools/src/modes/transform.ts

```diff
@@ -341,59 +341,115 @@ class Transformer {
     const options = optionsResult.result;
     const id = options.id ?? generateMsgIdFromAstNode(template, tag === 'html');
 
-    const sourceExpressions = new Map<string, ts.Expression>();
-    if (ts.isTemplateExpression(template)) {
-      for (const span of template.templateSpans) {
-        // TODO(aomarks) Support less brittle/more readable placeholder keys.
-        const key = this.sourceFile.text.slice(
-          span.expression.pos,
-          span.expression.end
-        );
-        sourceExpressions.set(key, span.expression);
-      }
-    }
-
-    // If translations are available, replace the source template from the
-    // second argument with the corresponding translation.
     if (this.translations !== undefined) {
       const translation = this.translations.get(id);
       if (translation !== undefined) {
+        // If translations are available, replace the source template from the
+        // second argument with the corresponding translation.
+
+        // Maps from <translation absolute expression index> to
+        // <[source placeholder index, placeholder-relative expression index]>.
+        const transExprToSourcePosition = new Map<number, [number, number]>();
+
+        // Maps from <source placeholder index> to <the number of expressions in
+        // that placeholder>.
+        const placeholderExpressionCounts = new Map<number, number>();
+
+        // The absolute position of each expression within the translated
+        // message.
+        let absTransExprIdx = 0;
+
+        // Maps source placeholder to their index.
+        const placeholdersByIndex = new Map<number, Placeholder>();
+        for (let i = 0, phIdx = 0; i < contents.length; i++) {
+          const content = contents[i];
+          if (typeof content === 'object') {
+            placeholdersByIndex.set(phIdx++, content);
+          }
+        }
+
         const templateLiteralBody = translation.contents
-          .map((content) =>
-            typeof content === 'string'
-              ? escapeTextContentToEmbedInTemplateLiteral(content)
-              : content.untranslatable
-          )
+          .map((content) => {
+            if (typeof content === 'string') {
+              return escapeTextContentToEmbedInTemplateLiteral(content);
+            }
+            if (content.index === undefined) {
+              throw new Error(
```

Comment:
I think this will currently throw on a transform when the input is XLB, since we won't have indexes in those files. I think that means https://github.com/lit/lit/issues/1897 needs to be fixed as part of this PR too or else a release would break all XLB use cases.

(To be honest, as far as I know there is only one customer of XLB, and I believe they are on a much older version anyway -- so it's unlikely anybody would be hit by this. However we should still not introduce the breakage [and side note maybe reach out to them to see what their status is!]).

Reply:
Just to clarify, this will need to be a breaking change for existing XLB files, but we should at least make sure it works for fresh builds.

Existing XLB files could then be migrated manually. (I think if we had a lot of XLB users, I would say we should provide a backwards-compatibility function that tries to make a best guess at the index from the contents of the `<ph>` tag when the index is missing, but I don't think we need it given the small [1] number of users).